### DISCUSSION TITLE
Solving broken builds when updating WPGraphQL

### DISCUSSION THREAD

**Discussion by moonmeister** [wp-graphql/wp-graphql#1461] [Ideas]
## Problem
I've seen a consistent issue with maintaining WPGraphQL on a production site. This: If I update WPGraphQL in WP my site build is broken until I'm able to update the front-end code. If I can use a dev version of WP to update WPGraphQL then I can update the front-end without affecting prod builds. But, eventually there will come a moment where I need to update WPGraphQL in prod and then merge the front-end changes. Even if that's only a couple minutes that's time where builds can be broken. On small static sites this isn't a huge problem, on large sites this may break publishing and previews for that time period (or if it's not statically generated, it might actually break the front-end of any site/client). 

## Possible solutions

Would it be possible...To allow installing multiple versions of WPGraphQL side-by-side. Not sure the specifics of this, but this would allow old versions of the site to continue working against older versions of WPGraphQL and new versions to to the new WPGraphQL API. 

## Things I've been thinking about:

1. Once WPGraphQL is stable the commonality of this issue is reduced significantly. That said it's still possible and potentially problematic when it does happen. 
2. It is possible to solve this at an ops level, but unless your the NYTimes, most people aren't running WP at the kind of scale to make that reasonable.
3. If someone were to scale using WPGraphQL to multiple clients, where they might not always be on the latest code (cached PWA, mobile, desktop, etc...), this would allow a long term solution to handling legacy code that hasn't been updated. 
4. This is not needed for V1, or not the initial release, but if implementation would be a breaking change maybe we think about building in the basics to allow it to be non-breaking in the future when implemented. 

@jasonbahl You may have a better less complicated solution but I've seen this my own site (I'm one person with a static site so not a big issue but it's a real problem with the Gatsby site. Kyle G and I were talking about it today so I thought I'd mention it. 🤷‍♂️

**Comment by seagyn** [comment]
Maybe an option would be to somehow cache the schema so requests still get the same response for a few minutes whilst the backend and frontend get back into sync.

Another option would be to try and handle schema updates on the frontend. An example is when we went to `0.10.x` and `author` and `featuredImage` etc now had a new shape you could go `if post.featuredImage.node ...  else if post.featuredImage ...` to handle both cases.

I do think that as soon as we hit 1.0 that this problem will be gone until 2.0 but then we can plan ahead of time for that.

I personally believe that this might be a lot of overhead that semver mostly solves.

**Comment by moonmeister** [comment]
> this might be a lot of overhead that semver mostly solves.

Very possible. I don't think it does entirely, so I thought I'd bring it up for discussion.

**Comment by seagyn** [comment]
Another option could be to allow a way to run two different versions on different endpoints so something like `/wpgraphql/0.10` `/wpgraphql/0.12` that way requests are version locked as well.

**Comment by moonmeister** [comment]
@seagyn This crossed my mind. I'm curious if it'd be possible to do this with headers instead of routes. Example: you add a `X-WPGRAPHQL-VERSION=^3.x` header to your request and then the `/graphql` endpoint handles routing it to use the best version. Possible benefit being you don't have to target different endpoints, you can just include the header in the request. If there's not a correct match for `^3.x` it can always fall back to the version 4 or 2 you do have installed and at least try to resolve instead of just getting a 404.  

I'm really curious what Jason has to say, he probable can think of a hundred things that affect this idea that may make it difficult or impossible.

**Comment by jasonbahl** [comment]
I don't think it's that straight forward unless we want to take on a LOT of technical debt. 

I did outline a "Seamless Upgrade Path" plan in the 1.0 release plan here: https://github.com/wp-graphql/wp-graphql/issues/1357

Where for some breaking changes at least, we could follow a 3-release approach:

- Release new feature under new name (consumers notified to update to use aliases)
- Release where old feature flips to new shape
- Release where old feature is fully removed and new feature takes over old name

---

There's also this library that looks interesting: https://github.com/ef-eng/graphql-query-rewrite

We could explore some way for folks to be able to alias old queries in some way. This will allow for WPGraphQL to move forward without maintaining gobs of tech debt, but allow each project the ability to safely move forward even if/when WPGraphQL core does make breaking changes.

**Comment by jasonbahl** [comment]
I think there's also a lot more room for tooling we can explore. Things like persisted query logging, so folks can track the query operations being made to their servers and when Schema changes are introduced they can identify which queries would break.

**Comment by moonmeister** [comment]
I knew Jason would have a highly intelligent answer for me. Thanks, sounds like it's on your mind and it'll get solved when it needs to. Feel free to close this issue, or whatever.

**Comment by jasonbahl** [comment]
@moonmeister I'll probably leave it open to get more discussion. I don't think it's a solved thing by any means yet.

**Comment by jasonbahl** [comment]
@moonmeister I'm transferring this to a discussion as it seems like a more open-ended topic right now. I think it's a good conversation to have that might stem many different ideas. It's for sure not a "solved" problem in the WPGraphQL space or even broader GraphQL space.
